src/tp_engine.cpp

```diff
@@ -159,23 +432,162 @@ void TPEngine::GrowSpareEdgeData(const size_t new_edge_spare_count) {
+            const Node *left_child_node, const Node *right_child_node) {
+          Bitset edge_bitset =
+              SBNMaps::PCSPBitsetOf(leaf_count, sister_node, false, focal_node, false,
+                                    left_child_node, false, right_child_node, false);
+          const auto edge_idx = AtWithDefault(edge_indexer, edge_bitset, default_index);
+          if (edge_idx != default_index) {
+            function_on_tree_node_by_edge(EdgeId(edge_idx), edge_bitset, tree, tree_id,
+                                          focal_node);
+          }
+        },
+        true);
+    tree_id++;
+  }
+}
+
+void TPEngine::SetChoiceMapByTakingFirst(const RootedTreeCollection &tree_collection,
+                                         const BitsetSizeMap &edge_indexer) {
+  input_tree_count_ = tree_collection.TreeCount();
+  const size_t tree_id_max = GetInputTreeCount();
+  tree_source_.resize(GetEdgeCount(), tree_id_max);
+  std::fill(tree_source_.begin(), tree_source_.end(), tree_id_max);
+  // Set tree source map for each edge in DAG.
+  auto set_tree_source = [tree_id_max, this](
+                             const EdgeId edge_idx, const Bitset &edge_bitset,
+                             const RootedTree &tree, const size_t tree_id,
+                             const Node *focal_node) {
+    if (tree_source_[edge_idx.value_] == tree_id_max) {
+      tree_source_[edge_idx.value_] = tree_id;
+    }
+  };
+  FunctionOverRootedTreeCollection(set_tree_source, tree_collection, edge_indexer);
+  // Set tree source map for rootsplit edges from the best tree.
+  const auto root_node = dag_.GetDAGNode(dag_.GetDAGRootNodeId());
+  for (const auto rootsplit_node_id : dag_.GetRootsplitNodeIds()) {
+    const auto rootsplit_node = dag_.GetDAGNode(rootsplit_node_id);
+    const auto rootsplit_edge_id = dag_.GetEdgeIdx(root_node.Id(), rootsplit_node.Id());
+    size_t best_tree_source = tree_id_max;
+    dag_.IterateOverLeafwardEdges(
+        rootsplit_node, [this, &rootsplit_node, &rootsplit_edge_id, &best_tree_source](
+                            bool is_edge_on_left, SubsplitDAGNode child_node) {
+          const auto edge_id = dag_.GetEdgeIdx(rootsplit_node.Id(), child_node.Id());
+          if (best_tree_source > tree_source_[edge_id.value_]) {
+            best_tree_source = tree_source_[edge_id.value_];
+            tree_source_[rootsplit_edge_id.value_] = best_tree_source;
+          }
+        });
+  }
+  // Build choice map from tree source.
+  for (EdgeId edge_id = 0; edge_id < GetEdgeCount(); edge_id++) {
+    const auto edge = dag_.GetDAGEdge(edge_id);
+    const auto parent = dag_.GetDAGNode(edge.GetParent());
+    const auto child = dag_.GetDAGNode(edge.GetChild());
+    const auto focal_clade = edge.GetSubsplitClade();
+    const auto sister_clade = Bitset::Opposite(focal_clade);
+    size_t best_tree_id;
+    // Get parent EdgeChoice.
+    best_tree_id = tree_id_max;
```

Comment:
A lot of code repetition here for the following 4 cases. Can this not get factored up into a little subroutine?

Reply:
Refactored.